src/main/asciidoc/java/index.adoc

```diff
@@ -0,0 +1,741 @@
+rules to make this possible:
+
+1. the API must be described as a set of Java interfaces, classes are not permitted
+2. nested interfaces are not permitted
+3. All interfaces to have generation performed on them must be annotated with the `io.vertx.codegen.annotations.VertxGen` annotation
+4. fluent methods (methods which return a reference to this) must be annotated with the `io.vertx.codegen.annotations.Fluent` annotation
+5. methods where the return value must be cached in the API shim must be annotated with the `io.vertx.codegen.annotations.CacheReturn` annotation
+6. only certain types are allowed as parameter or return value types for any API methods
+7. custom enums should be annotated with `@VertxGen`, although this is not mandatory to allow the usage of existing Java enums
+
+An API type can be generic or declare generic methods, type parameters must be unbounded, e.g
+`<N extends Number>` is forbidden.
+
+In the perspective of codegen, Java types can be categories as follow:
+
+. _basic_ type : any primitive/boxed type or `java.lang.String`
+. _json_ type : `io.vertx.core.json.JsonObject` or `io.vertx.core.json.JsonArray`
+. _api_ type : any type annotated with `io.vertx.codegen.annotations.VertxGen`
+. _data object_ type : any type annotated with `io.vertx.codegen.annotations.DataObject`
+. _enum_ type : any Java enum
+. _collection_ type : `java.util.List<V>`, `java.util.Set<V>` or `java.util.Map<String, V>`
+
+Parameterized types are supported but wildcards are not, that is the following type arguments declarations
+are forbidden:
+
+- `Foo<?>`
+- `Foo<? extends Number>`
+- `Foo<? super Number>`
+
+Parameterized types are only supported for _api_ generic types and _collection_ types.
+
+Type variables are allowed and carry a special meaning: a type variable is a dynamic form of a _basic_ type and
+_json_ type.
+
+==== Inheritance
+
+_api_ type can extend other _api_ types.
+
+An _api_ type can either be *concrete* or *abstract*, such information is importan for languages not
+supporting multiple class inheritance like Groovy:
+
+- _api_ types annotated with `link:../../apidocs/io/vertx/codegen/annotations/VertxGen.html[@VertxGen]``(concrete = false)` are meant to be
+extended by *concrete* interfaces an can inherit from *abstract* interfaces only.
+- _api_ types annotated with `link:../../apidocs/io/vertx/codegen/annotations/VertxGen.html[@VertxGen]` or `link:../../apidocs/io/vertx/codegen/annotations/VertxGen.html[@VertxGen]``(concrete = true)`
+are implemented directly by Vertx and can inherit at most one other *concrete* interface and any *abstract* interface
+
+==== Method return type
+
+The following return types are allowed:
+
+. `void` type
+. any _basic_ type
+. any _api_ type or parameterized _api_ type having type variable parameters
+. any _json_ type
+. the `java.lang.Throwable` type
+. any _enum_ type
+. any _data object_ type
+. an unbounded type variable
```

Comment:
ah ok, unbounded in the context of type variable and it should rather be "unbounded type parameter" https://docs.oracle.com/javase/tutorial/java/generics/bounded.html . will correct and add link